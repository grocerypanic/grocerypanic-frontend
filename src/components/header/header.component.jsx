import React from "react";

import StopIcon from "@material-ui/icons/Stop";
import AddIcon from "@material-ui/icons/Add";
import HomeIcon from "@material-ui/icons/Home";
import StoreIcon from "@material-ui/icons/Store";
import KitchenIcon from "@material-ui/icons/Kitchen";

import PropTypes from "prop-types";
import { withRouter } from "react-router-dom";

import { Navbar, Nav, Spinner } from "react-bootstrap";
import {
  NavContainer,
  IconContainer,
  OffIconContainer,
  MenuContainer,
} from "./header.styles";

import Routes from "../../configuration/routes";

const Header = ({ history, title, create, transaction }) => {
  React.useEffect(() => {}, []);

  const navigate = (route) => {
    if (transaction) return;
    history.push(route);
  };

  return (
    <NavContainer collapseOnSelect variant="dark" fixed="top">
      <Navbar.Brand href="#home">Panic: {title}</Navbar.Brand>
      <Nav className="mr-auto"></Nav>
      <Nav>
        <Nav.Item className="nav-link" onClick={() => navigate(Routes.root)}>
          <MenuContainer
            route={Routes.root}
            history={history}
            data-testid="home-icon"
          >
            <HomeIcon />
          </MenuContainer>
        </Nav.Item>
      </Nav>
      <Nav>
        <Nav.Item className="nav-link" onClick={() => navigate(Routes.stores)}>
          <MenuContainer
            route={Routes.stores}
            history={history}
            data-testid="stores-icon"
          >
            <StoreIcon />
          </MenuContainer>
        </Nav.Item>
      </Nav>
      <Nav>
        <Nav.Item className="nav-link" onClick={() => navigate(Routes.shelves)}>
          <MenuContainer
            route={Routes.shelves}
            history={history}
            data-testid="shelves-icon"
          >
            <KitchenIcon />
          </MenuContainer>
        </Nav.Item>
      </Nav>
      <Nav>
        {create ? (
          <Nav.Item className="nav-link" onClick={create}>
            <IconContainer>
              {transaction ? (
                <Spinner size="sm" animation="border" />
              ) : (
                <AddIcon className="AddIcon" />
              )}
            </IconContainer>
          </Nav.Item>
        ) : (
          <Nav.Item className="nav-link">
            <OffIconContainer>
              <StopIcon />
            </OffIconContainer>
          </Nav.Item>
        )}
      </Nav>
    </NavContainer>
  );
};

export default withRouter(Header);

Header.propTypes = {
  create: PropTypes.func,
  title: PropTypes.string.isRequired,
  transaction: PropTypes.bool.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
};
