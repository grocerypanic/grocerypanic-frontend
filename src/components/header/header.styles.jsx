import styled from "styled-components";
import { Navbar, Nav } from "react-bootstrap";
import { primary } from "../../configuration/theme";

export const NavContainer = styled(Navbar)`
  background-color: ${primary};
`;
