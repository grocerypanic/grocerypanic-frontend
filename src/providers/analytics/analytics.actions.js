export const AnalyticsActions = {
  TestAction: { category: "Test", action: "Submitted a test action" },
  LoginError: { category: "Error", action: "Unable to login with backend" },
};
