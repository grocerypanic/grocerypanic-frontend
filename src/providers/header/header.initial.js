const InitialValues = {
  headerSettings: {
    title: "",
    create: null,
    transaction: false,
    disableNav: false,
    signIn: false,
  },
  updateHeader: null,
};

export default InitialValues;
