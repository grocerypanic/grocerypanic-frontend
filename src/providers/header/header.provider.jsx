import React, { useState, createContext } from "react";
import InitialState from "./header.initial";

export const HeaderContext = createContext({ ...InitialState });

const HeaderProvider = ({ children }) => {
  const [headerSettings, setHeaderSettings] = useState(
    InitialState.headerSettings
  );

  const updateHeader = (newSettings) => {
    setHeaderSettings(newSettings);
  };

  return (
    <HeaderContext.Provider
      value={{
        headerSettings,
        updateHeader,
      }}
    >
      {children}
    </HeaderContext.Provider>
  );
};

export default HeaderProvider;
