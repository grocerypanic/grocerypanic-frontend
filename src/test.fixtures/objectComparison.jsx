export const propCount = (hash, length) => {
  expect(Object.keys(hash).length).toBe(length);
};
