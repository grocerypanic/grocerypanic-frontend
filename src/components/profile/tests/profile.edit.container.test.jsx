import { render, act } from "@testing-library/react";
import React from "react";
import { AnalyticsActions } from "../../../providers/analytics/analytics.actions";
import { AnalyticsContext } from "../../../providers/analytics/analytics.provider";
import ApiActions from "../../../providers/api/api.actions";
import mockTimezoneHook from "../../../providers/api/timezone/tests/timezone.hook.mock";
import useTimezones from "../../../providers/api/timezone/timezone.hook";
import mockProfileHook from "../../../providers/api/user/tests/user.hook.mock";
import useProfile from "../../../providers/api/user/user.hook";
import initialHeaderSettings from "../../../providers/header/header.initial";
import { HeaderContext } from "../../../providers/header/header.provider";
import useSocialLogin from "../../../providers/social/social.hook";
import mockSocialHook from "../../../providers/social/tests/social.hook.mock";
import { propCount } from "../../../test.fixtures/objectComparison";
import ErrorHandler from "../../error-handler/error-handler.component";
import HoldingPattern from "../../holding-pattern/holding-pattern.component";
import ProfileEditContainer from "../profile.edit.container";
import ProfileForm from "../profile.edit.form.component";

jest.mock("../profile.edit.form.component");
jest.mock("../../../providers/social/social.hook");
jest.mock("../../../providers/api/user/user.hook");
jest.mock("../../../providers/api/timezone/timezone.hook");
jest.mock("../../error-handler/error-handler.component");
jest.mock("../../holding-pattern/holding-pattern.component");

ProfileForm.mockImplementation(() => <div>MockForm</div>);
ErrorHandler.mockImplementation(({ children }) => children);
HoldingPattern.mockImplementation(({ children }) => children);

const mockAnalyticsContext = {
  initialized: true,
  event: jest.fn(),
  setup: true,
};

const mockHeaderContext = {
  ...initialHeaderSettings,
  updateHeader: jest.fn(),
};

const props = {
  headerTitle: "mockHeaderTitle",
  title: "mockTitle",
  helpText: "mockHelpText",
};

let utils;
let currentSocialHook;
let currentProfileHook;
let currentTimezoneHook;
let currentProps;

describe("Setup Environment", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    currentProps = { ...props };
    currentSocialHook = mockSocialHook();
    currentProfileHook = mockProfileHook();
    currentTimezoneHook = mockTimezoneHook();
  });

  const renderHelper = (currentProps, command = render) => {
    utils = command(
      <AnalyticsContext.Provider value={mockAnalyticsContext}>
        <HeaderContext.Provider value={mockHeaderContext}>
          <ProfileEditContainer {...currentProps} />
        </HeaderContext.Provider>
      </AnalyticsContext.Provider>
    );
  };

  const setupHooks = () => {
    useSocialLogin.mockImplementation(() => currentSocialHook);
    useProfile.mockImplementation(() => currentProfileHook);
    useTimezones.mockImplementation(() => currentTimezoneHook);
  };

  describe("outside of an error", () => {
    beforeEach(() => {
      currentProfileHook.profile.user.fail = false;
    });

    describe("with a valid user profile loaded", () => {
      beforeEach(() => {
        currentProfileHook.profile.user.inventory = ["mockData"];
      });

      describe("outside of a transaction", () => {
        beforeEach(() => {
          currentProfileHook.profile.user.transaction = false;
          setupHooks();
          renderHelper(currentProps);
        });

        it("should match the snapshot on file (styles)", () => {
          expect(utils.container).toMatchSnapshot();
        });

        it("renders, calls the expected hooks", () => {
          expect(currentProfileHook.profile.getProfile).toBeCalledTimes(1);
          expect(currentTimezoneHook.timezones.getTimezones).toBeCalledTimes(1);
        });

        it("renders, should call ErrorHandler with the correct params", () => {
          expect(ErrorHandler).toHaveBeenCalledTimes(1);
          const call = ErrorHandler.mock.calls[0][0];
          propCount(call, 5);
          expect(call.condition).toBeFalsy();
          expect(call.clearError).toBe(currentProfileHook.profile.clearErrors);
          expect(call.eventMessage).toBe(AnalyticsActions.ApiError);
          expect(call.messageTranslationKey).toBe("Profile.ApiError");
          expect(call.children).not.toBeNull();
        });

        it("renders, should call HoldingPattern with the correct params", () => {
          expect(HoldingPattern).toHaveBeenCalledTimes(1);
          const call = HoldingPattern.mock.calls[0][0];
          propCount(call, 2);
          expect(call.condition).toBeTruthy();
          expect(call.children).not.toBeNull();
        });

        it("renders, should call updateHeader with the correct params", () => {
          expect(mockHeaderContext.updateHeader).toHaveBeenCalledTimes(1);
          expect(mockHeaderContext.updateHeader).toBeCalledWith({
            title: currentProps.headerTitle,
            create: null,
            transaction: false,
            disableNav: false,
          });
        });

        it("renders, should call ProfileEditFrom with the correct params", () => {
          expect(ProfileForm).toHaveBeenCalledTimes(1);
          const call = ProfileForm.mock.calls[0][0];
          propCount(call, 6);
          expect(call.user).toBe(currentProfileHook.profile.user.inventory);
          expect(call.title).toBe(currentProps.title);
          expect(call.helpText).toBe(currentProps.helpText);
          expect(call.transaction).toBe(
            currentProfileHook.profile.user.transaction
          );
          expect(call.handleSave).toBeInstanceOf(Function);
          expect(call.timezones).toBe(
            currentTimezoneHook.timezones.timezones.inventory
          );
        });

        it("renders, saves profile data correctly", async () => {
          expect(ProfileForm).toHaveBeenCalledTimes(1);
          const call = ProfileForm.mock.calls[0][0];
          const save = call.handleSave;
          const mockObject = { id: 0, name: "mockObect" };

          await act(async () => save(mockObject));

          expect(mockAnalyticsContext.event).toBeCalledWith(
            AnalyticsActions.ProfileModified
          );
          expect(currentProfileHook.profile.updateProfile).toBeCalledWith(
            mockObject
          );
        });
      });

      describe("during a user transaction", () => {
        beforeEach(() => {
          currentProfileHook.profile.user.inventory = [];
          currentProfileHook.profile.user.transaction = true;
          currentTimezoneHook.timezones.timezones.transaction = false;
          setupHooks();
          renderHelper(currentProps);
        });

        it("renders, calls the expected hooks", () => {
          expect(currentProfileHook.profile.getProfile).toBeCalledTimes(1);
          expect(currentTimezoneHook.timezones.getTimezones).toBeCalledTimes(1);
        });

        it("renders, should call ErrorHandler with the correct params", () => {
          expect(ErrorHandler).toHaveBeenCalledTimes(1);
          const call = ErrorHandler.mock.calls[0][0];
          propCount(call, 5);
          expect(call.condition).toBeFalsy();
          expect(call.clearError).toBe(currentProfileHook.profile.clearErrors);
          expect(call.eventMessage).toBe(AnalyticsActions.ApiError);
          expect(call.messageTranslationKey).toBe("Profile.ApiError");
          expect(call.children).not.toBeNull();
        });

        it("renders, should call HoldingPattern with the correct params", () => {
          expect(HoldingPattern).toHaveBeenCalledTimes(1);
          const call = HoldingPattern.mock.calls[0][0];
          propCount(call, 2);
          expect(call.condition).toBeTruthy();
          expect(call.children).not.toBeNull();
        });

        it("renders, should call updateHeader with the correct params", () => {
          expect(mockHeaderContext.updateHeader).toHaveBeenCalledTimes(1);
          expect(mockHeaderContext.updateHeader).toBeCalledWith({
            title: currentProps.headerTitle,
            create: null,
            transaction: true,
            disableNav: false,
          });
        });

        it("renders, should call ProfileEditFrom with the correct params", () => {
          expect(ProfileForm).toHaveBeenCalledTimes(1);
          const call = ProfileForm.mock.calls[0][0];
          propCount(call, 6);
          expect(call.user).toBe(currentProfileHook.profile.user.inventory);
          expect(call.title).toBe(currentProps.title);
          expect(call.helpText).toBe(currentProps.helpText);
          expect(call.transaction).toBe(
            currentProfileHook.profile.user.transaction
          );
          expect(call.handleSave).toBeInstanceOf(Function);
          expect(call.timezones).toBe(
            currentTimezoneHook.timezones.timezones.inventory
          );
        });

        it("renders, does not save profile data when asked", async () => {
          expect(ProfileForm).toHaveBeenCalledTimes(1);
          const call = ProfileForm.mock.calls[0][0];
          const save = call.handleSave;
          const mockObject = { id: 0, name: "mockObect" };

          await act(async () => save(mockObject));
          expect(mockAnalyticsContext.event).toBeCalledTimes(0);
          expect(currentProfileHook.profile.updateProfile).toBeCalledTimes(0);
        });
      });

      describe("during a timezone transaction", () => {
        beforeEach(() => {
          currentTimezoneHook.timezones.timezones.inventory = [];
          currentTimezoneHook.timezones.timezones.transaction = true;
          currentProfileHook.profile.user.transaction = false;
          setupHooks();
          renderHelper(currentProps);
        });

        it("renders, calls the expected hooks", () => {
          expect(currentProfileHook.profile.getProfile).toBeCalledTimes(1);
          expect(currentTimezoneHook.timezones.getTimezones).toBeCalledTimes(1);
        });

        it("renders, should call ErrorHandler with the correct params", () => {
          expect(ErrorHandler).toHaveBeenCalledTimes(1);
          const call = ErrorHandler.mock.calls[0][0];
          propCount(call, 5);
          expect(call.condition).toBeFalsy();
          expect(call.clearError).toBe(currentProfileHook.profile.clearErrors);
          expect(call.eventMessage).toBe(AnalyticsActions.ApiError);
          expect(call.messageTranslationKey).toBe("Profile.ApiError");
          expect(call.children).not.toBeNull();
        });

        it("renders, should call HoldingPattern with the correct params", () => {
          expect(HoldingPattern).toHaveBeenCalledTimes(1);
          const call = HoldingPattern.mock.calls[0][0];
          propCount(call, 2);
          expect(call.condition).toBeTruthy();
          expect(call.children).not.toBeNull();
        });

        it("renders, should call updateHeader with the correct params", () => {
          expect(mockHeaderContext.updateHeader).toHaveBeenCalledTimes(1);
          expect(mockHeaderContext.updateHeader).toBeCalledWith({
            title: currentProps.headerTitle,
            create: null,
            transaction: true,
            disableNav: false,
          });
        });

        it("renders, should call ProfileEditFrom with the correct params", () => {
          expect(ProfileForm).toHaveBeenCalledTimes(1);
          const call = ProfileForm.mock.calls[0][0];
          propCount(call, 6);
          expect(call.user).toBe(currentProfileHook.profile.user.inventory);
          expect(call.title).toBe(currentProps.title);
          expect(call.helpText).toBe(currentProps.helpText);
          expect(call.transaction).toBe(
            currentProfileHook.profile.user.transaction
          );
          expect(call.handleSave).toBeInstanceOf(Function);
          expect(call.timezones).toBe(
            currentTimezoneHook.timezones.timezones.inventory
          );
        });

        it("renders, does not save profile data when asked", async () => {
          expect(ProfileForm).toHaveBeenCalledTimes(1);
          const call = ProfileForm.mock.calls[0][0];
          const save = call.handleSave;
          const mockObject = { id: 0, name: "mockObject" };

          await act(async () => save(mockObject));
          expect(mockAnalyticsContext.event).toBeCalledTimes(0);
          expect(currentProfileHook.profile.updateProfile).toBeCalledTimes(0);
        });
      });
    });

    describe("with timezones already loaded", () => {
      beforeEach(() => {
        currentProfileHook.profile.user.transaction = false;
        currentTimezoneHook.timezones.timezones.inventory = [
          { id: 1, name: "fake timezone" },
        ];
        setupHooks();
        renderHelper(currentProps);
      });

      it("renders, does not refetch the timezones", () => {
        expect(currentProfileHook.profile.getProfile).toBeCalledTimes(1);
        expect(currentTimezoneHook.timezones.getTimezones).toBeCalledTimes(0);
      });
    });

    describe("without a valid user profile loaded", () => {
      beforeEach(() => {
        currentProfileHook.profile.user.inventory = [];
        currentProfileHook.profile.user.transaction = false;
        setupHooks();
        renderHelper(currentProps);
      });

      it("renders, should call ErrorHandler with the correct params", () => {
        expect(ErrorHandler).toHaveBeenCalledTimes(1);
        const call = ErrorHandler.mock.calls[0][0];
        propCount(call, 5);
        expect(call.condition).toBeFalsy();
        expect(call.clearError).toBe(currentProfileHook.profile.clearErrors);
        expect(call.eventMessage).toBe(AnalyticsActions.ApiError);
        expect(call.messageTranslationKey).toBe("Profile.ApiError");
        expect(call.children).not.toBeNull();
      });

      it("renders, should call HoldingPattern with the correct params", () => {
        expect(HoldingPattern).toHaveBeenCalledTimes(1);
        const call = HoldingPattern.mock.calls[0][0];
        propCount(call, 2);
        expect(call.condition).toBeTruthy();
        expect(call.children).not.toBeNull();
      });
    });
  });

  describe("during an error", () => {
    describe("during an user api transaction failure", () => {
      beforeEach(() => {
        currentProfileHook.profile.user.fail = true;
        currentProfileHook.profile.user.transaction = false;
        setupHooks();
        renderHelper(currentProps);
      });

      it("renders, should call ErrorHandler with the correct params", () => {
        expect(ErrorHandler).toHaveBeenCalledTimes(1);
        const call = ErrorHandler.mock.calls[0][0];
        propCount(call, 5);
        expect(call.condition).toBeTruthy();
        expect(call.clearError).toBe(currentProfileHook.profile.clearErrors);
        expect(call.eventMessage).toBe(AnalyticsActions.ApiError);
        expect(call.messageTranslationKey).toBe("Profile.ApiError");
        expect(call.children).not.toBeNull();
      });
    });

    describe("during an timezone api transaction failure", () => {
      beforeEach(() => {
        currentTimezoneHook.timezones.timezones.fail = true;
        currentTimezoneHook.timezones.timezones.transaction = false;
        setupHooks();
        renderHelper(currentProps);
      });

      it("renders, should call ErrorHandler with the correct params", () => {
        expect(ErrorHandler).toHaveBeenCalledTimes(1);
        const call = ErrorHandler.mock.calls[0][0];
        propCount(call, 5);
        expect(call.condition).toBeTruthy();
        expect(call.clearError).toBe(currentProfileHook.profile.clearErrors);
        expect(call.eventMessage).toBe(AnalyticsActions.ApiError);
        expect(call.messageTranslationKey).toBe("Profile.ApiError");
        expect(call.children).not.toBeNull();
      });
    });

    describe("during a user authentication failure", () => {
      beforeEach(() => {
        currentProfileHook.profile.user.fail = true;
        currentProfileHook.profile.user.errorMessage = ApiActions.FailureAuth;
        setupHooks();
        renderHelper(currentProps);
      });

      it("renders, should call ErrorHandler with the correct params", () => {
        expect(currentSocialHook.social.expiredAuth).toHaveBeenCalledTimes(1);
      });
    });

    describe("during a timezones authentication failure", () => {
      beforeEach(() => {
        currentTimezoneHook.timezones.timezones.fail = true;
        currentTimezoneHook.timezones.timezones.errorMessage =
          ApiActions.FailureAuth;
        setupHooks();
        renderHelper(currentProps);
      });

      it("renders, should call ErrorHandler with the correct params", () => {
        expect(currentSocialHook.social.expiredAuth).toHaveBeenCalledTimes(1);
      });
    });
  });
});
