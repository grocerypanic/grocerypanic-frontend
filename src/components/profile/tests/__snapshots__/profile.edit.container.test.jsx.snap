// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Setup Environment outside of an error with a valid user profile loaded outside of a transaction should match the snapshot on file (styles) 1`] = `
<div>
  <div>
    MockForm
  </div>
</div>
`;
